File: LeetCodeTasks/23_merge-k-sorted-lists/solutions/java/solution.java
import java.util.PriorityQueue;


class Solution {

    public ListNode mergeKLists(ListNode[] lists) {

        if (lists == null || lists.length == 0) return null;

        

        PriorityQueue<ListNode> heap = new PriorityQueue<>((a, b) -> a.val - b.val);

        

        // Initialize the heap with the first node of each list if it is not empty

        for (ListNode node : lists) {

            if (node != null) {

                heap.add(node);

            }

        }

        

        ListNode dummy = new ListNode(0);

        ListNode current = dummy;

        

        while (!heap.isEmpty()) {

            ListNode minNode = heap.poll();

            current.next = minNode;

            current = current.next;

            if (minNode.next != null) {

                heap.add(minNode.next);

            }

        }

        

        return dummy.next;

    }

}

File: LeetCodeTasks/23_merge-k-sorted-lists/solutions/c/solution.c
#include <stdlib.h>


struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {

    struct ListNode *head = NULL, **tail = &head;

    while (l1 && l2) {

        if (l1->val < l2->val) {

            *tail = l1;

            l1 = l1->next;

        } else {

            *tail = l2;

            l2 = l2->next;

        }

        tail = &(*tail)->next;

    }

    *tail = l1 ? l1 : l2;

    return head;

}



struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {

    if (listsSize == 0) return NULL;

    int interval = 1;

    while (interval < listsSize) {

        for (int i = 0; i < listsSize - interval; i += 2 * interval) {

            lists[i] = mergeTwoLists(lists[i], lists[i + interval]);

        }

        interval *= 2;

    }

    return lists[0];

}

